cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapperWrapper.java

```diff
@@ -154,11 +163,36 @@ public boolean nextKeyValue() throws IOException, InterruptedException {
         }
         return result;
       }
+
+      @Override
+      public InputSplit getInputSplit() {
+        InputSplit inputSplit = super.getInputSplit();
+        if (inputSplit instanceof TaggedInputSplit) {
+          // expose the delegate InputSplit to the user
+          inputSplit = ((TaggedInputSplit) inputSplit).getInputSplit();
+        }
+        return inputSplit;
+      }
+
+      @Override
+      public Class<? extends InputFormat<?, ?>> getInputFormatClass() throws ClassNotFoundException {
+        InputSplit inputSplit = super.getInputSplit();
+        if (inputSplit instanceof TaggedInputSplit) {
+          // expose the delegate InputFormat to the user
+          return ((TaggedInputSplit) inputSplit).getInputFormatClass();
+        }
+        return super.getInputFormatClass();
+      }
     };
     return flushingContext;
   }
 
-  private Mapper createMapperInstance(ClassLoader classLoader, String userMapper) {
+  private Mapper createMapperInstance(ClassLoader classLoader, String userMapper, Context context) {
+    if (DelegatingMapper.class.getName().equals(userMapper)) {
```

Comment:
Why it is in the `Mapper` wrapper logic? Shouldn't it be the `DelegatingMapper` responsibility? Also delegation of the context for resolving to the actual Input name and input format should be part of the DelegatingMapper as well, isn't it?